### PR TITLE
Handle subordinate-subordinate relations in subordinate filtering

### PR DESCRIPTION
**Pull request by babbageclunk**
## Description of change
In fixing [bug 1686696](https://bugs.launchpad.net/juju/+bug/1686696) I changed the uniter API WatchUnitRelations to restrict container-scoped relations to those where the other end was the app for the principal unit. Then in the fixes for the related [bug 1699050](https://bugs.launchpad.net/juju/+bug/1699050) an upgrade step was added to make the same change in the database, and the uniter API was changed to prevent invalid relation units entering scope.

This excluded relations between subordinate applications (like ntp and nrpe), which meant that upgrading a model with sub-sub relations would result in the uniters repeatedly crashing as they tried to remove a relation state directory with files in it (because that sub-sub relation had been incorrectly filtered out).

Change the watcher, upgrade step and uniter API EnterScope method to allow sub-sub relations through.

## QA steps

* Bootstrap a 2.1.3 controller and add a model.
* Deploy the following bundle:
```
series: xenial
machines:
  '0':
    series: xenial
  '1':
    series: xenial
services:
  min:
    charm: cs:ubuntu-10
    num_units: 1
    to:
      - "0"
  min2:
    charm: cs:ubuntu-10
    num_units: 1
    to:
      - "1"
  nrpe:
    charm: cs:nrpe-23
  ntp:
    charm: cs:ntp-18
relations:
- - min:juju-info
  - nrpe:general-info
- - min2:juju-info
  - nrpe:general-info
- - ntp:nrpe-external-master
  - nrpe:nrpe-external-master
- - ntp:juju-info
  - min:juju-info
- - ntp:juju-info
  - min2:juju-info
```
* Observe in the DB that the min-ntp, min2-ntp, min-nrpe and min2-nrpe relations all have an incorrect unitcount of 3, but the ntp-nrpe relation unitcount is 4 (which is right).
* Upgrade the controller and model from 2.1.3 to the version containing this PR.
* Check in the DB that the incorrect unitcounts have been changed to 2, but the ntp-nrpe relation is still 4.
* Check that there aren't repeated errors in the log from the uniter crashing and restarting in a loop.

## Bug reference

Fixes https://bugs.launchpad.net/juju/+bug/1700450

### CODE REVIEW

File: state/relationunit.go [hunk 1]
```diff
@@ -356,19 +356,42 @@ func (ru *RelationUnit) Valid() (bool, error) {
...
+	var otherAppName string
+	for _, ep := range ru.relation.Endpoints() {
+		if ep.ApplicationName != appName {
+			otherAppName = ep.ApplicationName
+		}
+	}
+	if otherAppName == "" {
+		return false, errors.Errorf("couldn't find other endpoint")
+	}
+	app, err := ru.st.Application(otherAppName)
```
Comment (mjs) [inline state/relationunit.go]: Maybe rename `app` to `otherApp` so that it links up with `otherAppName`.

File: state/relationunit.go [hunk 2]
```diff
@@ -356,19 +356,42 @@ func (ru *RelationUnit) Valid() (bool, error) {
...
-	appName, err := names.UnitApplication(name)
+	pName, _ := unit.PrincipalName()
+	principalAppName, err := names.UnitApplication(pName)
 	if err != nil {
 		return false, errors.Trace(err)
 	}
-	_, err = ru.relation.Endpoint(appName)
+	// If the other application is a principal, only allow it if it's in the relation.
+	_, err = ru.relation.Endpoint(principalAppName)
 	return err == nil, nil
```
Comment (mjs) [inline state/relationunit.go]: I would have thought that if `err` is anything other than "not found" then `false, errors.Trace(err)` should be returned.

Comment (babbageclunk) [inline state/relationunit.go]: Yeah, I'll tighten that up - at the moment the only error `Endpoint` can return is the one indicating that the app passed in isn't an endpoint, but this code shouldn't assume that.

File: state/relationunit.go (hunk 1 above)
Comment (babbageclunk) [inline state/relationunit.go]: Will do.

**Comment by babbageclunk** [comment]
Intermittent failure reported here: https://bugs.launchpad.net/juju/+bug/1702215

!!build!!

**Comment by babbageclunk** [comment]
$$merge$$

**Comment by Veebers** [comment]
$$merge$$

**Comment by babbageclunk** [comment]
$$merge$$

(PR to skip serverSuite.TestClosesFromStatePool: #7594)

**Pull request merged**